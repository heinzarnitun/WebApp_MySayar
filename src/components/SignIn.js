import React, { useState, useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { login } from '../services/authService';

import axios from 'axios';
import API_URL from '../api';

const SignIn = () => {
  const [userData, setUserData] = useState({ email: '', password: '' });
  const [error, setError] = useState(''); // To handle error messages
  const navigate = useNavigate();
  const [success, setSuccess] = useState('');


  const handleChange = (e) => {
    setUserData({ ...userData, [e.target.name]: e.target.value });
  };

  const handleSubmit = async (e) => {
  e.preventDefault();
  setError('');

  try {
    // Fetch CSRF token before login
    await axios.get(`${API_URL}/sanctum/csrf-cookie`);  

    // Send login request
    const response = await axios.post(`${API_URL}/api/login`, userData);

    // Log the response to see its structure
    console.log(response.data);

    // Check if response contains the expected properties
    if (response.data && response.data.token && response.data.user) {
      const userId = response.data.user.id;

      // Store token and userId in localStorage
      localStorage.setItem('authToken', response.data.token);
      localStorage.setItem('user_id', userId);
      localStorage.setItem('username', response.data.user.name);
      console.log('Stored user_id:', localStorage.getItem('user_id'));

      setSuccess('Login successful! Redirecting...');

      // Redirect to dashboard after login
      setTimeout(() => {
        navigate('/dashboard');
        window.location.reload();
      }, 1000);
    } else {
      setError('Login failed: Incorrect credentials or server error');
    }

  } catch (error) {
    setError('Login failed: Incorrect credentials or server error');
    console.error(error);  // Log error details for debugging
  }
};

 return (
    <div className="container mt-5  mb-5">
      <div className="row justify-content-center">
        <div className="col-md-6 col-lg-4">
          <div className="card shadow">
            <div className="card-body p-4">
              {/* Header */}
              <div className="text-center mb-4">
                <h2 className="fw-bold">Sign In</h2>
                <p className="text-muted">Access your account</p>
              </div>

              {/* Messages */}
              {error && (
                <div className="alert alert-danger" role="alert">
                  {error}
                </div>
              )}
              {success && (
                <div className="alert alert-success" role="alert">
                  {success}
                </div>
              )}

              {/* Form */}
              <form onSubmit={handleSubmit}>
                <div className="mb-3">
                  <label htmlFor="email" className="form-label">Email</label>
                  <input
                    type="email"
                    className="form-control"
                    id="email"
                    name="email"
                    placeholder="your@email.com"
                    onChange={handleChange}
                    required
                  />
                </div>

                <div className="mb-4">
                  <label htmlFor="password" className="form-label">Password</label>
                  <input
                    type="password"
                    className="form-control"
                    id="password"
                    name="password"
                    placeholder="••••••••"
                    onChange={handleChange}
                    required
                  />
                </div>

                <button 
                  type="submit" 
                  className="btn btn-primary w-100 py-2 fw-bold"
                >
                  Sign In
                </button>
              </form>

              {/* Footer Links */}
              <div className="mt-3 text-center">
                <p className="text-muted">
                  Don't have an account? <a href="/sign-up" className="text-decoration-none">Sign up</a>
                </p>
                <p className="small">
                
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default SignIn;
