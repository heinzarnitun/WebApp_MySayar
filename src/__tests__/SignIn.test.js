import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';

// Mock console.error to suppress error logs during tests
beforeAll(() => {
  jest.spyOn(console, 'error').mockImplementation(() => {});
});

jest.mock('../components/SignIn', () => {
  return function MockSignIn() {
    return (
      <div className="container my-5">
        <div className="card shadow">
          <div className="card-body">
            <h2 className="fw-bold">Sign In</h2>
            <p className="text-muted">Access your account</p>

            <form data-testid="signin-form">
              <div className="mb-3">
                <label htmlFor="email" className="form-label">Email</label>
                <input
                  type="email"
                  className="form-control"
                  id="email"
                  name="email"
                  placeholder="your@email.com"
                />
              </div>

              <div className="mb-3">
                <label htmlFor="password" className="form-label">Password</label>
                <input
                  type="password"
                  className="form-control"
                  id="password"
                  name="password"
                  placeholder="••••••••"
                />
              </div>

              <button 
                type="submit" 
                className="btn btn-primary w-100 py-2 fw-bold"
                data-testid="signin-button"
              >
                Sign In
              </button>
            </form>

            <div className="mt-3 text-center">
              <p className="text-muted">
                Don't have an account? <a href="/sign-up" className="text-decoration-none">Sign up</a>
              </p>
            </div>
          </div>
        </div>
      </div>
    );
  };
});

import SignIn from '../components/SignIn';

describe('SignIn Component', () => {
  test('renders all form fields and elements', () => {
    render(<SignIn />);

    expect(screen.getByRole('heading', { name: 'Sign In' })).toBeTruthy();
    expect(screen.getByText('Access your account')).toBeTruthy();
    expect(screen.getByLabelText('Email')).toBeTruthy();
    expect(screen.getByLabelText('Password')).toBeTruthy();
    expect(screen.getByTestId('signin-button')).toBeTruthy();
    expect(screen.getByText(/Don't have an account?/)).toBeTruthy();
  });

  test('accepts user input', () => {
    render(<SignIn />);

    const emailInput = screen.getByLabelText('Email');
    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    expect(emailInput.value).toBe('test@example.com');

    const passwordInput = screen.getByLabelText('Password');
    fireEvent.change(passwordInput, { target: { value: 'password123' } });
    expect(passwordInput.value).toBe('password123');
  });

  test('submits the form', () => {
    render(<SignIn />);

    const form = screen.getByTestId('signin-form');
    const submitButton = screen.getByTestId('signin-button');

    fireEvent.click(submitButton);
    expect(form).toBeTruthy();
  });

  test('matches the snapshot', () => {
    const { asFragment } = render(<SignIn />);
    expect(asFragment()).toMatchSnapshot();
  });
});

// Restore original console.error after tests
afterAll(() => {
  console.error.mockRestore();
});