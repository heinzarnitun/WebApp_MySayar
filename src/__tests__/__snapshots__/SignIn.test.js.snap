// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container my-5"
  >
    <div
      class="card shadow"
    >
      <div
        class="card-body"
      >
        <h2
          class="fw-bold"
        >
          Sign In
        </h2>
        <p
          class="text-muted"
        >
          Access your account
        </p>
        <form
          data-testid="signin-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              placeholder="your@email.com"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              placeholder="••••••••"
              type="password"
            />
          </div>
          <button
            class="btn btn-primary w-100 py-2 fw-bold"
            data-testid="signin-button"
            type="submit"
          >
            Sign In
          </button>
        </form>
        <div
          class="mt-3 text-center"
        >
          <p
            class="text-muted"
          >
            Don't have an account? 
            <a
              class="text-decoration-none"
              href="/sign-up"
            >
              Sign up
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
